Swift auth update

chmouel opened the pull request.

chmouel committed:
Update swift token middleware.

Update swift middleware which would delegate authentication to
token_auth and handle ACL, container_sync and s3.

chmouel committed:
Remove executable bit.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -1,7 +1,4 @@
-#!/usr/bin/env python
-# vim: tabstop=4 shiftwidth=4 softtabstop=4
```
leave this in pls

chmouel committed:
Readd shbang and vim header.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env python
```
not this ;)

chmouel replied:
ah! i actually looked over the other middlewares and there is some who has
it... will add a commit that remove from all of them.

chmouel committed:
Remove python header shbang

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +16,181 @@
-"""
-TOKEN-BASED AUTH MIDDLEWARE FOR SWIFT
+from swift.common.utils import get_logger, split_path, get_remote_client
+from swift.common.middleware.acl import clean_acl, parse_acl, referrer_allowed
```
if you're going to be touching these files anyway, please update them to conform to the usual code style: https://github.com/openstack/nova/blob/master/HACKING.rst

basically in this case it means from swift.common.middleware import acl

and then everywhere you use clean_acl and such, you'll say acl.clean_acl and so forth

termie commented:
man this diff is really hard to read for some reason... anyway, still going through

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-    * use knowledge of tenant, admin status, and container acls to authorize
+class SwiftAuth(object):
+    """
+    Keystone to Swift authorization system.
```
this should be on the line above, like
    """Keystone to Swift authorization system.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+    Keystone to Swift authorization system.
 
-"""
+    Add to your pipeline in proxy-server.conf, such as::
```
could we mention somewhere that this is expected to be run on the swfit side of the world, not in the keystone side? it will be confusing if people look at this and wonder why it doesn't look like keystone middleware

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-        """
+        self.conf = conf
+        self.logger = get_logger(conf, log_route='keystone')
+        self.reseller_prefix = conf.get('reseller_prefix', 'AUTH').strip()
```
it seems like these defaults should be defined somewhere so that conf.get() automatically returns the default if it doesn't exist, isn't that a feature somewhere?

chmouel replied:
> it seems like these defaults should be defined somewhere so that
> conf.get() automatically returns the default if it doesn't exist, isn't
> that a feature somewhere?

Are you talking about that kind of stuff :

https://github.com/openstack/keystone/blob/master/keystone/middleware/auth_token.py#L142

https://github.com/openstack/keystone/blob/master/keystone/middleware/auth_token.py#L177

https://github.com/openstack/keystone/blob/master/keystone/middleware/auth_token.py#L217

It doesn't seem that the other middleware in ksl are doing such things as
well so I didn't bother doing it, can you point me to an example?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+            conf.get('keystone_tenant_user_admin', "false").lower() in \
+            ('true', 't', '1', 'on', 'yes', 'y')
+        self.allowed_sync_hosts = [h.strip()
+            for h in conf.get('allowed_sync_hosts', '127.0.0.1').split(',')
```
could you assign conf.get(...).split(',') to a variable first so that this line isn't so long?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        self.keystone_swift_operator_roles = \
+            conf.get('keystone_swift_operator_roles', 'Admin, SwiftOperator')
+        self.keystone_tenant_user_admin = \
+            conf.get('keystone_tenant_user_admin', "false").lower() in \
```
it seems like this config value is a weird name for a boolean, seems like it should have an "_is_" in there somewhere.

also, what does this mean if it is true?

chmouel replied:
| > +        self.keystone_tenant_user_admin = \

> > -            conf.get('keystone_tenant_user_admin', "false").lower() in \
> >   it seems like this config value is a weird name for a boolean, seems like
> >   it should have an "_is_" in there somewhere.
> >   also, what does this mean?

This is to mimic a RAX Cloud Files authentication, when this option is set
if the tenant_name == user_name then it grant automatically Admin access to
this user_name for this account/tenant.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        identity = self._keystone_identity(environ)
+
+        if not identity:
+            #TODO: non authenticated access allow via refer
```
TODO's should be formatted as # TODO(chmouel): message

and i don't understand what this todo means, could you expand it a bit?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        return self.app(environ, start_response)
+
+    def _keystone_identity(self, environ):
+        """ Extract the identity from the Keystone auth component """
```
No spaces around the docstring, period at the end please

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+
+    def _keystone_identity(self, environ):
+        """ Extract the identity from the Keystone auth component """
+        if (environ.get('HTTP_X_IDENTITY_STATUS') != 'Confirmed'):
```
outer parens not necessary

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        if (environ.get('HTTP_X_IDENTITY_STATUS') != 'Confirmed'):
+            return None
+        roles = []
+        if ('HTTP_X_ROLE' in environ):
```
no outer parens necessary

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+            return None
+        roles = []
+        if ('HTTP_X_ROLE' in environ):
+            roles = environ.get('HTTP_X_ROLE').split(',')
```
you don't really need to use the .get() syntax if you've already checked for the existence of the key... but not a big deal

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-        env['swift.clean_acl'] = self.authorize_via_acl
-        return self.app(env, start_response)
+    def _reseller_check(self, account, tenant_id):
+        """ Check reseller prefix """
```
No spaces around message and a period pls.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-        if account:
-            return self._authorize_anon_account(req, account)
+        if not self._reseller_check(account, tenant[0]):
+            self.logger.debug('tenant mismatch')
```
could you add the two tenants that didn't match to the logging message

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+
+        user_groups = env_identity.get('roles', [])
+
+        # If user is in the swift operator group then make the owner of it.
```
i don't understand this message, could you make it more clear what it is saying?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        user_groups = env_identity.get('roles', [])
+
+        # If user is in the swift operator group then make the owner of it.
+        for _group in self.keystone_swift_operator_roles.split(','):
```
it isn't very common to prefix variables with underscores, usually it menas that it is an unused variable (people will do stuff like:

for k, _unused in dict.iteritems():
   ... 

is there a reason why you named this this way? if you are just trying to avoid a name conflict the usual way is to put the underscore at the end (group_) but that is only usually done for reserved words

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+            _group = _group.strip()
+            if  _group in user_groups:
+                self.logger.debug(
+                    "User in group: %s allow to manage this account" % \
```
you don't need to escape newlines while inside parens, also this might be easier broken up like so:

self.logger.debug("User in groups: %s allow"
                                 " to manage this account % (group))

... with appropriate indenting, not sure how to do that in comment field

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        # If user is of the same name of the tenant then make owner of it.
+        user = env_identity.get('user', '')
+        if self.keystone_tenant_user_admin and user == tenant[1]:
+            self.logger.debug("user: %s == %s tenant and option "\
```
while in parens you don't need to escape newlines

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
 
-        return self._authorize_anon_toplevel(req)
+        # Allow container sync
+        if (req.environ.get('swift_sync_key') and
```
i like to put the "and" parts at the beginning of the next line so it is clear that there is a continuation happening

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+            return None
 
-    def _authorize_anon_object(self, req, account, container, obj):
+        # Check if Referrer allow it
```
does this mean "check if referrer is allowed" ?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+
+        # Allow ACL at individual user level (tenant:user format)
+        if '%s:%s' % (tenant[0], user) in groups:
+            self.logger.debug('user explicitly allowed in ACL authorizing')
```
could you include which user is allowed in the message

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+            if user_group in groups:
+                self.logger.debug('user in group which is allowed in' \
+                        ' ACL: %s authorizing' % (user_group))
+                return None
```
all these return None should probably just be "return" ?

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-        resp = conn.getresponse()
-        data = resp.read()
-        conn.close()
+        # last but not least retun deny
```
probably don't need this comment

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
-        return identity
+    def denied_response(self, req):
+        """
+        Returns a standard WSGI response callable with the status of 403 or 401
```
please add a one-line summary as the first line of the docstring

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,181 @@
+        depending on whether the REMOTE_USER is set or not.
+        """
+        if req.remote_user:
+            return HTTPForbidden(request=req)
```
these will probably turn into webob.exc.HTTPForbidden

chmouel committed:
address termie's comment on review.

- Mostly code style.
- Change the variable keystone_tenant_user_admin to
  keystone_tenant_user_is_admin.

chmouel committed:
Import swift imports with swift_ prefix.

termie commented on keystone/middleware/swift_auth.py:
```diff
@@ -16,220 +15,188 @@
-"""
-TOKEN-BASED AUTH MIDDLEWARE FOR SWIFT
+from swift.common import utils as swift_utils
+from swift.common.middleware import acl as swift_acl
```
this is fine, but for the record i mostly cared about the utils one just because it is the same name as keystone utils (and many other "utils" across the projects)

chmouel committed:
Rename variables dropping keystone_ prefix.

This is more verbosity when it was not needed.

chmouel committed:
More missing period for docs and syntax.

chmouel committed:
Change Admin to admin.

As this seem to be the de facto role for the admin user now.
